src/cart/crt_init.c

```diff
@@ -52,6 +52,16 @@ crt_lib_init(void)
 	crt_gdata.cg_iv_inline_limit = 19456; /* 19KB */
 }
 
+void
+crt_reset_afterfork(void)
```

Comment:
`crt_reset` or `cart_lib_reset` would be better names here, the code doesn't know or care when you call it.

Should there be any locking here?  There appears to be no checking of thread safety at all.

There are at least two mutexes being wiped here that should really be finalised before being wiped but it's probably better to change them to be static variables on their own and use static initialisers for them rather than calling _INIT and _FINI on them.

Reply:
Thank you! This function will be called right after fork() inside the newly spawned child process. I think there is no need to use lock. There is only a single thread in a new process.
It would be more clear with "_after_fork". I am not aware of any other situations we need to call this function at this moment. If we do need to call it somewhere else, we could rename it later.

Reply:
You are right. We could move this function to src/cart/crt_internal.h as you suggested. Thank you!

Reply:
I am not sure this approach is safe:

if we assume that fork() can happen at any time after crt_init() was called, then we need to do a clean finalization of all cart, and not just reset some variables.

For example, assume an original process allocated few contexts and initialized a provider.
After that a fork() happens and this function gets eventually called.
Even though a forked child can in theory now re-init cart, you now have lost all references to cart/mercury contexts and other data. 

Is there a reason why in daos lib you cant simply call crt_finalize() followed by crt_init() again?

Reply:
Once fork() is called, a child process is created and proceed from the fork point. The parent process may still need to utilize the existing context, so we can NOT call crt_finalize() at this point. It is the parent process's responsibility to destroy the context when the parent process ends. The child process inherited the memory from its parent process due to fork. We do need to reset the global variables and create a fresh context inside the child process. The parent and child process have separate memory.

Reply:
I understand now what you are trying to do, but i think it will require a bit more changes.

crt_gdata contains number of elements that get allocated either at init or during cart usage, and would need to be free-ed first before setting gdata_init_flag = 0 and memsetting, in order to not leak memory.

For example crt_gdata->cg_opc_map will be allocated based on registered RPCs in the parent.

There are also other potential SW things you would need to deallocate and reset.
For example crt_group.c contains global cart group list and associated lock.

  19 /* global CRT group list */
  20 D_LIST_HEAD(crt_grp_list);


It also raises a question what to do with cart contexts if any are already allocated.
As you mentioned we cannot destroy the context, as it is parents responsibility, but at the same time we should destroy all SW-related parts of the cart context to not leak memory.

the main concern here is that cart contexts contain rank-to-hg_handle caches and these would be bad to leak after fork().
Similarly crt_gdata contains rank-to-uri cache that also needs to be nuked to not leak memory.

Reply:
why would there be leaked memory on the child if one just reset the init flag?
all memory inherited for the child on fork is copy-on-write, no?

Reply:
If i understand it correctly with a copy-on-write, upon fork() the pages will get marked as read-only, parent and child both point to the same pages and a write by either a parent or child will result in page fault, followed by a copy of a particular page.

So with this, a parent writing to crt_gdata right after the fork() will result in a child getting its own copy (parents original), and when reset that copy will be lost.
Also since copy-on-write is on per-page boundaries, writing to something unrelated could result in a copy.

Do you think its safe in this case?

Reply:
@frostedcmos . crt_gdata is set zero inside child process here. Do we need to re-initialize "crt_grp_list" too here? Are there any other global variables needed to reset?

Reply:
> If i understand it correctly with a copy-on-write, upon fork() the pages will get marked as read-only, parent and child both point to the same pages and a write by either a parent or child will result in page fault, followed by a copy of a particular page.

yea indeed. i did not consider that the parent can change this global data struct.